### PR TITLE
feat: pagination: adds pagination component

### PR DESCRIPTION
**Pull request by dkilgore-eightfold**
Jira Ticket:
[ENG-2790: [Octuple] Pagination Component Suite](https://eightfoldai.atlassian.net/browse/ENG-2790)

### CODE REVIEW

File: src/components/Pagination/Pagination.tsx [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+            return pages;
+        };
+
+        const jumpToPage = (page: number): void => {
+            const oldVal: number = _currentPage;
+
+            setCurrentPage(page);
+
+            if (oldVal !== _currentPage) {
+                onCurrentChange && onCurrentChange(_currentPage);
```
Comment (ychhabra-eightfold) [inline src/components/Pagination/Pagination.tsx:101]: same as above

File: src/components/Pagination/Pagination.tsx [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            }
+        };
+
+        const handleCurrentChange = (val: number): void => {
+            const oldVal: number = _currentPage;
+
+            setCurrentPage(val);
+
+            if (oldVal !== val) {
+                onCurrentChange && onCurrentChange(val);
```
Comment (ychhabra-eightfold) [inline src/components/Pagination/Pagination.tsx:111]: same as above

**Review by ychhabra-eightfold — CHANGES_REQUESTED** [review]

**Comment by ychhabra-eightfold** [comment]
How to use the goto input doesnt seem to work for me, also the page selector doesnt seem to work sometimes

File: src/components/Pagination/Pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+                data-test-id={dataTestId}
+            >
+                {layout.includes(PaginationLayoutOptions.Sizes) && (
+                    <span className={styles.sizes} key="sizes">
+                        {/** TODO: Replace with Select component when available */}
+                        <Dropdown
+                            overlay={Overlay(_pageSizes)}
+                            onVisibleChange={setToggle}
+                        >
+                            <DefaultButton
```
[the remark this answers is not in the document]
Comment (dkilgore-eightfold) [reply src/components/Pagination/Pagination.tsx:176]: We can use props for these, will add.

**Comment by dkilgore-eightfold** [comment]
> How to use the goto input doesnt seem to work for me, also the page selector doesnt seem to work sometimes

For the Goto button to work you'll need the latest main - as there was a bug where onChange was doing nothing in Input. Regarding the page selector, this may be some other issue. Works for me on a few browsers, so may be some other issue related to Menu component

File: src/components/Pagination/Pagination.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (dkilgore-eightfold) [reply src/components/Pagination/Pagination.tsx:176]: That being said, there's no real need for aria-labels for buttons that only contain a count. Will add aria-label support for the icon buttons only.

File: src/components/Pagination/Pager.tsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        const getVisiblePageRange = useCallback((): Array<number> => {
+            const pagerCount: number = 7;
+            let array: number[] = [];
+
+            if (currentPage > 5 && !(currentPage < pageCount - 5)) {
+                const startPage = pageCount - (pagerCount - 2);
+
+                for (let i: number = startPage; i < pageCount; ++i) {
+                    array.push(i);
+                }
```
[the remark this answers is not in the document]
Comment (dkilgore-eightfold) [reply src/components/Pagination/Pager.tsx]: Not as readable, but will apply.

File: src/components/Pagination/Pager.tsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        { currentPage = 1, onCurrentChange, pageCount, ...rest },
+        ref: Ref<HTMLUListElement>
+    ) => {
+        const [_pagers, setPagers] = useState<number[]>([0]);
+        const [_quickNextActive, setQuickNextActive] = useState<boolean>(false);
+        const [_quickPreviousActive, setQuickPreviousActive] =
+            useState<boolean>(false);
+
+        const getVisiblePageRange = useCallback((): Array<number> => {
+            const pagerCount: number = 7;
```
[the remark this answers is not in the document]
Comment (dkilgore-eightfold) [reply src/components/Pagination/Pager.tsx]: Added comment, keeping scoped.

File: src/components/Pagination/Pager.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            let array: number[] = [];
+
+            if (currentPage > 5 && !(currentPage < pageCount - 5)) {
```
[the remark this answers is not in the document]
Comment (dkilgore-eightfold) [reply src/components/Pagination/Pager.tsx]: Moved these to a constant

File: src/components/Pagination/Pager.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkilgore-eightfold) [reply src/components/Pagination/Pager.tsx]: Keeping the legible loop for now.

**Review by ychhabra-eightfold — APPROVED** [review]

**Comment by ychhabra-eightfold** [comment]
LGTM, updaten snapshots before merging

**Pull request merged**